libraries/Microsoft.Bot.Builder.Azure/CosmosDbPartitionedStorage.cs

```diff
@@ -201,28 +204,47 @@ public async Task WriteAsync(IDictionary<string, object> changes, CancellationTo
-                {
-                    // if new item or * then insert or replace unconditionally
-                    await _container.UpsertItemAsync(
-                            documentChange,
-                            GetPartitionKey(documentChange.PartitionKey),
-                            cancellationToken: cancellationToken)
-                        .ConfigureAwait(false);
-                }
-                else if (etag.Length > 0)
+
+                try
                 {
-                    // if we have an etag, do opt. concurrency replace
-                    await _container.UpsertItemAsync(
-                            documentChange,
-                            GetPartitionKey(documentChange.PartitionKey),
-                            new ItemRequestOptions() { IfMatchEtag = etag, },
-                            cancellationToken)
-                        .ConfigureAwait(false);
+                    if (etag == null || etag == "*")
+                    {
+                        // if new item or * then insert or replace unconditionally
+                        await _container.UpsertItemAsync(
+                                documentChange,
+                                GetPartitionKey(documentChange.PartitionKey),
+                                cancellationToken: cancellationToken)
+                            .ConfigureAwait(false);
+                    }
+                    else if (etag.Length > 0)
+                    {
+                        // if we have an etag, do opt. concurrency replace
+                        await _container.UpsertItemAsync(
+                                documentChange,
+                                GetPartitionKey(documentChange.PartitionKey),
+                                new ItemRequestOptions() { IfMatchEtag = etag, },
+                                cancellationToken)
+                            .ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        throw new Exception("etag empty");
+                    }
                 }
-                else
+                catch (CosmosException ex)
                 {
-                    throw new Exception("etag empty");
+                    // This check could potentially be performed before even attempting to upsert the item
+                    // so that a request wouldn't be made to Cosmos if it's expected to fail.
+                    // However, performing the check here ensures that this custom exception is only thrown
+                    // if Cosmos returns an error first.
+                    // This way, the nesting limit is not imposed on the Bot Framework side
+                    // and no exception will be thrown if the limit is eventually changed on the Cosmos side.
+                    if (GetNestingErrorMessage(json) is string message)
+                    {
+                        throw new BotStateException(message, ex);
```

Comment:
Why create a custom exception? Can we just throw with a good message?

Reply:
Do you mean throw the base `Exception` type?